### ISSUE TITLE
Socket.close() not freeing resources after send/recv timeout

### ISSUE THREAD

**Issue by etyrrill** [zeromq/jeromq#448]
I am running into a problem where jeromq is not freeing resources if the Socket is unable to connect.  This is reproducible with the following code:

```
import org.zeromq.ZMQ;
import org.zeromq.ZMQ.Context;
import org.zeromq.ZMQ.Socket;

public class rrclient {
    public static void main(String[] args) {
        Context context = ZMQ.context(1);

        while (true) {
            try (Socket socket = context.socket(ZMQ.REQ)) {
                socket.connect("tcp://localhost:33333");
                socket.setSendTimeOut(1000);
                socket.setReceiveTimeOut(1000);
                socket.send("Hello");
                String reply = socket.recvStr(0);
                if (reply == null) {
                    System.out.println("TIMEOUT");
                } else {
                    System.out.println("Received reply [" + reply + "]");
                }
            } catch (Exception e) {
                e.printStackTrace();
                break;
            }
        }
        context.term();
    }
}
```

Simply run this code without a server listening on the other end, and the process will eventually consume all file handles on the machine.  I was able to reproduce this on linux and macos.

**Comment by etyrrill** [comment]
BTW tested this with 0.3.6, 0.4.0, and 0.4.2 and the problem exists with all three.

**Comment by daveyarwood** [comment]
It looks like this program creates sockets forever and never closes them. Is that intentional?

**Comment by etyrrill** [comment]
The example code is using "try with resources", introduced in java 7.  Any objects instantiated in the parens after the try must implement Closable, and close() is automatically called on the Closable(s) after the end of the try block regardless of if an exception is thrown in the try block.  It works similar to if you had added:

```
finally {
    socket.close();
}
```

after the catch.

**Comment by trevorbernard** [comment]
@etyrrill What's your Operating system? This could be platform specific.

**Comment by etyrrill** [comment]
The problem exists on linux (centos 7) and mac (osx 10.11)

**Comment by etyrrill** [comment]
Oh, and we've also seen this on Ubuntu 16.10 as well.

**Comment by trevorbernard** [comment]
Ok, sounds like it likely isn't a platform specific thing

**Comment by colini** [comment]
In 0MQ `socket.close()` is an asynchronous operation.  In a loop like that it's possible sockets are being created faster than they are cleaned up.

**Comment by etyrrill** [comment]
Take two minutes, run my test program, attach jvisualvm to it, and you can watch the socket count go up only, not down.  The socket, and associated resources seem to never be freed, even after an hour or more.

The only time I see the sockets freed is when I bring up a REP socket listening at the other end, though that only works if done before the machine runs out of file handles.

**Comment by trevorbernard** [comment]
@etyrrill Unfortunately, I can't recreate the memory leak. It's running stable on MacOS 10.12.5. [Yourkit screenshot](http://imgur.com/a/lfROq)

```$ java -version
java version "1.8.0_101"
Java(TM) SE Runtime Environment (build 1.8.0_101-b13)
Java HotSpot(TM) 64-Bit Server VM (build 25.101-b13, mixed mode)
```

**Comment by etyrrill** [comment]
Hi Trevor,

How long did you let it run?  On my mac it takes about 20 minutes before I get the following stack trace:
```
java.lang.IllegalStateException: EMFILE
	at zmq.Ctx.createSocket(Ctx.java:375)
	at org.zeromq.ZMQ$Socket.<init>(ZMQ.java:432)
	at org.zeromq.ZMQ$Context.socket(ZMQ.java:356)
	at rrclient.main(rrclient.java:11)
```
My java:
```
openjdk version "1.8.0_121"
OpenJDK Runtime Environment (build 1.8.0_121-b13)
OpenJDK 64-Bit Server VM (build 25.121-b13, mixed mode)
```

**Comment by etyrrill** [comment]
I took a video of the test program running side by side with jvisualvm showing the socket count rising until it gets the exception above https://youtu.be/MgDUDaJZXQQ.

**Comment by colini** [comment]
I can reproduce it on MacOS 10.12.6 with IntelliJ and Oracle java 1.8.0_51.  Jeromq 0.4.2.  It did take a while but I eventually saw the EMFILE error.

**Comment by fredoboulo** [comment]
Reproduced it on my machine as well.

Inserting that helps to speed up the failing process:
`
        Context context = ZMQ.context(1);
        context.setMaxSockets(10);
`

Setting this line as the last one in the try block fixes the situation:
`                socket.setLinger(0);
            }
            catch (Exception e) {
...
`

**Comment by daveyarwood** [comment]
It looks like ZContext automatically sets linger (to 0 by default) when destroying sockets, but ZMQ.Context does not, which could be part of the issue.

**Comment by fredoboulo** [comment]
I would not advocate for setting linger to 0, it would discard all messages not already sent to the network, regardless of the user's settings.

More and more, I think that the problem lies before closing.

Allow me to sum up the code assessing the problem:
- REQ is connected, with a session (standard case)
- as **immediate is true by default**, one pipe connecting socket and session is immediately created, but with no peer over the network.
- one message is sent to the pipe, and accumulated there waiting for a peer to appear.
- send returns true in every case, because message was successfully queued into the pipe.
- received is performed and times out.

Closing is then performed, with a linger = -1, so socket waits until all messages are sent through the network. **Forever** as there are no peer.

I would go for setting **immediate = false** for REQ, PUSH, DEALER, but this has an impact on existing code using the lib, so I would like to be rather careful with that (I did enough damage control in the last weeks).

[Here doc for setting the option](https://github.com/zeromq/libzmq/blob/master/doc/zmq_setsockopt.txt#L385). This was particularly confusing when rewriting to 4.1.7, libzmq seems to have [changed the name of the option](https://github.com/zeromq/libzmq/commit/8358d4e8328895456399f8c2ee8a0f273fa3f8dc#diff-8ee357e14391bd5746684fd9bbb1a071) (formerly DELAY_ATTACH_ON_CONNECT_COULD_THIS_BE_ANY_LONGER to quote a good mind), but not inverted the logic. Reading this passage tells that if this option is set to 1, sending shall wait until a connection is complete,  and default is 0 (false)
So my understanding is Immediate default 0 means shall not wait until a connection is complete, so Immediate=true if we want to keep the semantic. 

TLDR; I think the problem lies more in the fact that the message is sent with no blocking because immediate option is true by default, behavior that I preserved when bumping to 4.1.7.

**Comment by daveyarwood** [comment]
@etyrrill Does the previous post help?

**Comment by daveyarwood** [comment]
Closing due to inactivity. Feel free to reopen if this is still an issue and the post above doesn't address the problem.
